File: src/main/java/hu/frontrider/arcana/research/theory/CardExperiment.java
package hu.frontrider.arcana.research.theory;

import hu.frontrider.arcana.items.Rodent;
import hu.frontrider.arcana.util.AspectUtil;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.text.TextComponentTranslation;
import net.minecraftforge.fml.common.registry.GameRegistry;
import thaumcraft.api.ThaumcraftApiHelper;
import thaumcraft.api.research.theorycraft.ResearchTableData;
import thaumcraft.api.research.theorycraft.TheorycraftCard;

import static hu.frontrider.arcana.ThaumicArcana.MODID;


public class CardExperiment extends TheorycraftCard {
    @GameRegistry.ObjectHolder(MODID + ":rodent")
    static Item rodent = null;

    @Override
    public int getInspirationCost() {
        return 3;
    }

    @Override
    public String getLocalizedName() {
        return (new TextComponentTranslation("card.thaumic_arcana.experiment.name")).toString();
    }

    @Override
    public boolean[] getRequiredItemsConsumed() {
        return new boolean[]{false, true, true};
    }

    @Override
    public String getLocalizedText() {
        return (new TextComponentTranslation("card.thaumic_arcana.experiment.description")).toString();
    }

    @Override
    public ItemStack[] getRequiredItems() {
        return new ItemStack[]{
                new ItemStack(CardExperiment.rodent, 1),
                ThaumcraftApiHelper.makeCrystal(AspectUtil.getRandom()),
                ThaumcraftApiHelper.makeCrystal(AspectUtil.getRandom())
        };
    }

    @Override
    public boolean activate(EntityPlayer player, ResearchTableData data) {
        for (ItemStack item : player.inventory.mainInventory) {
            if (item.getItem() == rodent)
                if (!((Rodent) item.getItem()).isDead(item, player.getEntityWorld())) {
                    player.inventory.addItemStackToInventory(new ItemStack(CardExperiment.rodent, 1));
                    data.addTotal(this.getResearchCategory(), 25);
                    return true;
                }
        }
        return false;
    }

    @Override
    public String getResearchCategory() {
        return "BIOMANCY";
    }
}

File: src/main/java/hu/frontrider/arcana/research/theory/CardExperiment.kt
package hu.frontrider.arcana.research.theory

import hu.frontrider.arcana.ThaumicArcana.MODID
import hu.frontrider.arcana.items.Rodent
import hu.frontrider.arcana.util.AspectUtil
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.util.text.TextComponentTranslation
import net.minecraftforge.fml.common.registry.GameRegistry
import thaumcraft.api.ThaumcraftApiHelper
import thaumcraft.api.research.theorycraft.ResearchTableData
import thaumcraft.api.research.theorycraft.TheorycraftCard


class CardExperiment : TheorycraftCard() {

    override fun getInspirationCost(): Int {
        return 3
    }

    override fun getLocalizedName(): String {
        return TextComponentTranslation("card.thaumic_arcana.experiment.name").toString()
    }

    override fun getRequiredItemsConsumed(): BooleanArray {
        return booleanArrayOf(false, true, true)
    }

    override fun getLocalizedText(): String {
        return TextComponentTranslation("card.thaumic_arcana.experiment.description").toString()
    }

    override fun getRequiredItems(): Array<ItemStack> {
        return arrayOf(ItemStack(CardExperiment.rodent!!, 1), ThaumcraftApiHelper.makeCrystal(AspectUtil.random), ThaumcraftApiHelper.makeCrystal(AspectUtil.random))
    }

    override fun activate(player: EntityPlayer, data: ResearchTableData): Boolean {
        for (item in player.inventory.mainInventory) {
            if (item.item === rodent)
                if (((item.item is Rodent) && !Rodent.isDead(item, player.entityWorld))) {
                    player.inventory.addItemStackToInventory(ItemStack(CardExperiment.rodent!!, 1))
                    data.addTotal(this.researchCategory, 25)
                    item.shrink(1)
                    return true
                }
        }
        return false
    }

    override fun getResearchCategory(): String {
        return "BIOMANCY"
    }

    companion object {
        @GameRegistry.ObjectHolder("$MODID:rodent")
        internal var rodent: Item? = null
    }
}
